cpp/src/join/join_new.cu

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/join_new.hpp>
+
+namespace cudf {
+
+namespace join {
+
+namespace detail {
+
+  // TODO:
+  //  nested_join_indices      - my function/kernel
+  //  sort_merge_join_indices  - Kumar
+  //  hash_join_indices        - rework current hash mechanism
+  //  filter_join_indices      - new function/kernel
+  //
+  struct nested_loop_join {
+    VectorPair join_indices(table_view const& left,
+                            table_view const& right,
+                            std::vector<join_operation> const& primary_join_ops,
+                            std::vector<join_operation> const& secondary_join_ops,
+                            cudaStream_t stream) {
+
+      return nested_join_indices(left, right, primary_join_ops, stream);
+    }
+  };
+
+  struct sort_merge_join {
+    VectorPair operator()(table_view const& left,
+                          table_view const& right,
+                          std::vector<join_operation> const& primary_join_ops,
+                          std::vector<join_operation> const& secondary_join_ops,
+                          cudaStream_t stream) {
+
+      auto joined_indices = sort_merge_join_indices(left, right, primary_join_ops, stream);
+
+      if (secondary_join_ops.empty()) {
+        return joined_indices;
+      }
+
+      return filter_join_indices(left, right, joined_indices, secondary_join_ops, stream);
+    }
+  };
+
+  struct hash_join {
+    VectorPair operator()(table_view const& left,
+                          table_view const& right,
+                          std::vector<join_operation> const& primary_join_ops,
+                          std::vector<join_operation> const& secondary_join_ops,
+                          cudaStream_t stream) {
+
+      auto joined_indices = hash_join_indices(left, right, primary_join_ops, stream);
+
+      if (secondary_join_ops.empty()) {
+        return joined_indices;
+      }
+
+      return filter_join_indices(left, right, joined_indices, secondary_join_ops, stream);
+    }
+  };
+    
+  bool is_trivial_join(table_view const& left,
```

Comment:
Adaptation of current is_trivial_join to use primary_join_ops and secondary_join_ops